fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java

```diff
@@ -0,0 +1,193 @@
+            if (!predicate.isAnd()) {
+                context.setAnd(false);
+            }
+            RangePredicate childRange = child.accept(this, context);
+            context.setAnd(isAndOrigin);
+            if (childRange == null) {
+                if (!context.isAnd() || !predicate.isAnd()) {
+                    return null;
+                } else {
+                    if (!(child instanceof BinaryPredicateOperator)) {
+                        residualPredicates.add(child);
+                    }
+                    continue;
+                }
+            }
+            if (predicate.isOr()) {
+                if (childRange instanceof ColumnRangePredicate) {
+                    ColumnRangePredicate childColumnRange = childRange.cast();
+                    Optional<RangePredicate> rangePredicateOpt = findColumnRangePredicate(rangePredicates, childColumnRange);
+                    if (rangePredicateOpt.isPresent()) {
+                        ColumnRangePredicate matchedColumnRangePredicate = rangePredicateOpt.get().cast();
+                        ColumnRangePredicate newPredicate =
+                                matchedColumnRangePredicate.orRange(matchedColumnRangePredicate, childColumnRange);
+                        rangePredicates.remove(matchedColumnRangePredicate);
+                        if (!newPredicate.isUnbounded()) {
+                            rangePredicates.add(newPredicate);
+                        }
+                    } else {
+                        if (!((ColumnRangePredicate) childRange).isUnbounded()) {
+                            rangePredicates.add(childRange);
+                        }
+                    }
+                } else {
+                    rangePredicates.add(childRange);
+                }
+            } else if (predicate.isAnd()) {
+                if (childRange instanceof ColumnRangePredicate) {
+                    ColumnRangePredicate childColumnRange = childRange.cast();
+                    Optional<RangePredicate> rangePredicateOpt = findColumnRangePredicate(rangePredicates, childColumnRange);
+                    if (rangePredicateOpt.isPresent()) {
+                        ColumnRangePredicate newPredicate =
+                                ColumnRangePredicate.andRange(rangePredicateOpt.get().cast(), childColumnRange);
+                        rangePredicates.remove(rangePredicateOpt.get());
+                        rangePredicates.add(newPredicate);
+                    } else {
+                        if (!childColumnRange.isUnbounded()) {
+                            rangePredicates.add(childRange);
+                        }
+                    }
+
+                } else {
+                    rangePredicates.add(childRange);
+                }
+            } else if (predicate.isNot()) {
+                return null;
+            }
+        }
+        if (predicate.isAnd()) {
+            return new AndRangePredicate(rangePredicates);
```

Comment:
`rangePredicates ` may contain both or  / and  predicates?

Reply:
yes.